Add collection of ps_mem output to juju hosts.

fginther opened the pull request:
This creates a /var/log/ps_mem.txt file on each juju host machine (not on lxds). The logs are then collected via the bootstrap unit (machine 0) and the base-machine units in the inner environment. Collection looks like:
```
2017-01-24 19:56:53,238 INFO Collecting ps_mem output for all hosts including bootstrap
2017-01-24 19:56:55,364 INFO Collecting ps_mem output on unit 1
2017-01-24 19:57:00,300 INFO Collecting ps_mem output on unit 0
2017-01-24 19:57:07,115 INFO Collecting ps_mem output on unit 3
2017-01-24 19:57:13,722 INFO Collecting ps_mem output on unit 2
2017-01-24 19:57:25,716 INFO Collecting ps_mem output on unit 5
2017-01-24 19:57:31,989 INFO Collecting ps_mem output on unit 4
```

ps_mem.txt looks like:
```
 Private  +   Shared  =  RAM used   Swap used   Program

132.0 KiB +  53.0 KiB = 185.0 KiB         0.0 KiB     0.0 KiB   tee
192.0 KiB +  30.0 KiB = 222.0 KiB         0.0 KiB     0.0 KiB   acpid
  8.0 KiB + 268.0 KiB = 276.0 KiB       220.0 KiB   220.0 KiB   rabbitmq-server (2)
 80.0 KiB + 253.0 KiB = 333.0 KiB       108.0 KiB   144.0 KiB   inet_gethost (2)
...
  1.0 GiB +  47.7 MiB =   1.0 GiB        28.6 MiB   123.0 MiB   neutron-server (11)
  1.8 GiB +   1.8 MiB =   1.8 GiB       147.2 MiB   147.2 MiB   mongod (2)
  1.9 GiB +   4.2 MiB =   1.9 GiB         9.9 MiB     9.9 MiB   landscape-clien (8)
-------------------------------------------------------------
                         10.7 GiB       901.9 MiB     1.7 GiB
=============================================================
```

This example is stored under:
`landscape-0-inner-logs/base-machine-0/var/log/ps_mem.txt`


This was tested manually on a local deployment by running `collect-logs logs.tar.gz`.

fginther committed:
Add collection of ps_mem output to juju hosts.

blackboxsw commented on collect-logs:
```diff
@@ -267,18 +268,63 @@ def get_units(juju, status=None):
-    log.info("Collecting ps output on unit {}".format(unit.name))
-    ps_cmd = "ps fauxww | sudo tee /var/log/ps-fauxww.txt"
-    args = juju.ssh_args(unit, ps_cmd)
+def get_hosts(juju, status=None):
```
style nit. One more leading empty line before function defitions

fginther replied:
Fixed.

blackboxsw commented on collect-logs:
```diff
@@ -366,12 +412,16 @@ def collect_logs(juju):
     map(partial(_create_ps_output_file, juju), units)
 
+    log.info("Collecting ps_mem output for all hosts including bootstrap")
+    hosts = get_hosts(juju)
```
Can we perform the git clone here, and pass the ps_mem.py file path into _create_ps_mem_output_file function to limit our exposure to external github connectivity issues and avoid having to install git on all target units?

fginther replied:
Yes, the clone is now only done on the host executing `collect-logs`.

blackboxsw commented on collect-logs:
```diff
@@ -267,18 +268,63 @@ def get_units(juju, status=None):
+        return
+
+    # Fetch the ps_mem utility from the github repo
+    ps_mem_repo = "https://github.com/fginther/ps_mem.git"
```
Can we avoid duplicating this git clone on every JujuHost and instead juju.push_args(unit, ps_memfile, "/tmp/ps_mem.py"). It would also allow us greater control in install git in our ci environment instead of having to install git on every target machine.

fginther replied:
There is now an `upload_ps_mem()` which does the git clone on the host system executing `collect-logs`. This is called twice, once for the other host 0 and again to upload ps_mem onto the landscape-server/0 node for use by the inner model.

blackboxsw commented on collect-logs:
```diff
@@ -267,18 +268,63 @@ def get_units(juju, status=None):
+
+
+def _run_cmd(juju, unit, cmd, message):
+    """Helper method for running a command on a juju unit."""
```
docstirng s/juju unit/juju unit or host/
If feels like _run_cmd "message" parameter could be a generic description and used for both initial log message and failure messages.
Something like
def  _run_command(juju, unit, cmd, description="collect ps output"):
log.info("Performing '{}' on unit {}".format(description, unit.name))
try: 
  check_output(...)
except ...:
   log.error("Failed '{}' on unit '{}'.format(description, unit.name)).

Then we could drop the log.info above each _run_cmd call.

fginther replied:
I like the suggestion and I've updated `_run_cmd()`.

blackboxsw commented on test_collect-logs.py:
```diff
@@ -506,6 +556,19 @@ def test_get_units_failure(self):
 
+    def test_get_hosts_failure(self):
+        """
+        collect_logs() does not handle errors from get_hosts().
```
can be 1 line

fginther replied:
Kept these as multi-line as it matched the majority of the rest of the file (although it's a bit mixed).

blackboxsw commented on test_collect-logs.py:
```diff
@@ -569,6 +634,10 @@ def setUp(self):
             ]
         script.get_units.return_value = self.units[:]
+        self.hosts = [
+            script.JujuHost("0", "1.2.3.4.8"),
```
Interesting IP address

fginther replied:
Fixed

blackboxsw commented on test_collect-logs.py:
```diff
@@ -322,6 +323,10 @@ def setUp(self):
         script.get_units.return_value = self.units[:]
         script.get_bootstrap_ip.return_value = self.units[0].ip
+        self.hosts = [
+            script.JujuHost("0", "1.2.3.4.8"),
```
invalid IP.

fginther replied:
Fixed.

blackboxsw commented on test_collect-logs.py:
```diff
@@ -360,6 +365,27 @@ def test_success(self):
                                       env=None,
                                       ))
+        # for _create_ps_mem_output_file()
+        for host in self.hosts:
```
Since our unit test is only setting up a single host, maybe we can explicitly check that there is only one mocked juju ssh "0", cmd called.

fginther replied:
I've refactored the test to better exercise the new methods under a separate class. The test code to look for a long list of expected `subprocess.check_output` calls has been removed.

blackboxsw reviewed (changes requested):
Few minor changes to change where the ps_mem.py is cloned. I'll review again once those changes/responses are in place.

fginther committed:
Clone the git repo once on the collect-logs host.

Then upload it to each host unit which needs to execute it.
Refactor tests to exercise the added methods instead of watching
for the same subprocess calls in multiple places.

blackboxsw commented:
Good 1 fginther. Land away.
